apps/voting/app/src/vote-hooks.js

```diff
@@ -0,0 +1,80 @@
+import { useState, useEffect } from 'react'
+import { useAragonApi } from '@aragon/api-react'
+
+// when `params` are updated, call `fn` and pass the result
+function usePromiseResult(fn, params, memoParams, defaultValue) {
+  const [result, setResult] = useState(defaultValue)
+  useEffect(() => {
+    let cancelled = false
+    fn(...params).then(value => {
+      if (!cancelled) {
+        setResult(value)
+      }
+    })
+    return () => {
+      cancelled = true
+    }
+  }, memoParams)
+  return result
+}
+
+async function getUserBalance(vote, userAccount, tokenContract, tokenDecimals) {
+  if (!tokenContract || !userAccount) {
+    return -1
+  }
+
+  const balance = await tokenContract
+    .balanceOfAt(userAccount, vote.data.snapshotBlock)
+    .toPromise()
+
+  return Math.floor(parseInt(balance, 10) / Math.pow(10, tokenDecimals))
+}
+
+async function getCanVote(vote, userAccount, api) {
+  if (!vote) {
+    return false
+  }
+
+  // If the account is not present, we assume the account is not connected.
+  if (!userAccount) {
+    return vote.data.open
+  }
+
+  return api.call('canVote', vote.voteId, userAccount).toPromise()
+}
+
+async function getCanExecute(vote, api) {
```

Comment:
Technically, there's no need for both `getCanVote()` and `getCanExecute()` to be declared `async`

Reply:
They both return either promises or values, so `async` is needed to return a promise in every case, or am I missing something?

Also I think having `async` in front of a function returning a promise is a nice touch for readability (it doesn’t hurt performances since promises auto unwrap). It also ensures that no errors are thrown outside the returned promise.

Reply:
Ahh good point, didn't realize they were returning normal values too for the default case. And I do agree adding the `async` is a nice syntactic touch, let's keep doing that 👍.